### PR TITLE
curvefs/client: fix read failure in multi-mount scenario

### PR DESCRIPTION
**Pull request by baijiaruo**
### What problem does this PR solve?

Issue Number: #1240  

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 1]
```diff
@@ -361,19 +363,27 @@ int FileCacheManager::Read(uint64_t inodeId, uint64_t offset, uint64_t length,
...
-                    ::curve::common::UniqueLock lgGuard =
-                        inodeWrapper->GetUniqueLock();
-                    auto r = inodeWrapper->RefreshS3ChunkInfo();
-                    if (r != CURVEFS_ERROR::OK) {
-                        LOG(WARNING) << "refresh inode fail, ret:" << ret;
-                        return -1;
+                    if (retry < maxRetry) {
+                        // ret -2 refs s3obj not exist
+                        // clear inodecache && get again
+                        LOG(INFO) << "inode cache maybe steal, try to get "
```
Comment (wu-hanqing) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:373]: typo: steal -> stale ?

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 2]
```diff
@@ -361,19 +363,27 @@ int FileCacheManager::Read(uint64_t inodeId, uint64_t offset, uint64_t length,
...
+                            inodeWrapper->GetUniqueLock();
+                        auto r = inodeWrapper->RefreshS3ChunkInfo();
+                        if (r != CURVEFS_ERROR::OK) {
+                            LOG(WARNING) << "refresh inode fail, ret:" << ret;
+                            return -1;
+                        }
+                        bthread_usleep(retryIntervalMs * retry * 1000);
+                    } else {
+                        LOG(WARNING) << "retry is reach max retry:" << maxRetry;
+                        bthread_usleep(retryIntervalMs * maxRetry * 1000);
```
Comment (wu-hanqing) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:386]: if so, the `maxRetry` doesn't means the total number of retries, it means the max sleep intervals between retries.

File: curvefs/src/client/s3/client_s3_adaptor.h [hunk 3]
```diff
@@ -197,6 +197,14 @@ class S3ClientAdaptorImpl : public S3ClientAdaptor {
        diskCacheType_ = type;
     }
 
+    uint32_t GetMaxReadRetry() {
```
Comment (wu-hanqing) [inline curvefs/src/client/s3/client_s3_adaptor.h:200]: `uint32_t GetMaxReadRetry() const`

File: curvefs/src/client/s3/client_s3_adaptor.h [hunk 4]
```diff
@@ -197,6 +197,14 @@ class S3ClientAdaptorImpl : public S3ClientAdaptor {
...
+        return maxReadRetry_;
+    }
+
+    uint32_t GetReadRetryIntervalMs() {
```
Comment (wu-hanqing) [inline curvefs/src/client/s3/client_s3_adaptor.h:204]: `uint32_t GetReadRetryIntervalMs() const`

File: curvefs/conf/client.conf [hunk 5]
```diff
@@ -153,7 +153,8 @@ s3.throttle.bpsTotalMB=0
 s3.throttle.bpsReadMB=0
 s3.throttle.bpsWriteMB=0
 s3.useVirtualAddressing=false
-
+s3.maxReadRetry = 100
+s3.readRetryIntervalMs = 100
```
Comment (wu-hanqing) [inline curvefs/conf/client.conf:157]: add some explanations

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 6]
```diff
@@ -296,8 +296,10 @@ int FileCacheManager::Read(uint64_t inodeId, uint64_t offset, uint64_t length,
     }
 
     {
-        unsigned int maxRetry = 3;  // hardcode, fixme
-        unsigned int retry = 0;
+        uint32_t maxRetry =
+            s3ClientAdaptor_->GetMaxReadRetry();  // hardcode, fixme
```
Comment (ilixiaocui) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:300]: The default here should be infinite retries. It is recommended to use 0 to indicate infinite retries

File: curvefs/src/client/s3/client_s3_cache_manager.cpp (hunk 2 above)
Comment (ilixiaocui) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:386]: I have the same doubt.

File: curvefs/src/client/s3/client_s3_cache_manager.cpp [hunk 7]
```diff
@@ -361,19 +363,27 @@ int FileCacheManager::Read(uint64_t inodeId, uint64_t offset, uint64_t length,
             if (ret < 0) {
                 retry++;
                 responses.clear();
-                if (ret != -2 || retry == maxRetry) {
+                if (ret != -2) {
```
Comment (ilixiaocui) [inline curvefs/src/client/s3/client_s3_cache_manager.cpp:366]: what does ret==-2 mean?

File: curvefs/conf/client.conf (hunk 5 above)
Comment (baijiaruo) [reply curvefs/conf/client.conf:157]: > add some explanations
fixed

File: curvefs/src/client/s3/client_s3_adaptor.h (hunk 3 above)
Comment (baijiaruo) [reply curvefs/src/client/s3/client_s3_adaptor.h:200]: fixed

File: curvefs/src/client/s3/client_s3_adaptor.h (hunk 4 above)
Comment (baijiaruo) [reply curvefs/src/client/s3/client_s3_adaptor.h:204]: fixed

File: curvefs/src/client/s3/client_s3_adaptor.h [hunk 8]
```diff
@@ -234,6 +242,8 @@ class S3ClientAdaptorImpl : public S3ClientAdaptor {
     uint32_t chunkFlushThreads_;
     uint32_t memCacheNearfullRatio_;
     uint32_t throttleBaseSleepUs_;
+    uint32_t maxReadRetry_;
+    uint32_t readRetryIntervalMs_;
```
[the remark this answers is not in the document]
Comment (baijiaruo) [reply curvefs/src/client/s3/client_s3_adaptor.h:246]: I don't think there is a difference

File: curvefs/src/client/s3/client_s3_cache_manager.cpp (hunk 6 above)
Comment (baijiaruo) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:300]: maxRetry is changed to maxIntervalNum, which means the maximum number of intervals

File: curvefs/src/client/s3/client_s3_cache_manager.cpp (hunk 7 above)
Comment (baijiaruo) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:366]: This is the linux error code, which means ENOENT，indicating that the data cannot be read on s3

File: curvefs/src/client/s3/client_s3_cache_manager.cpp (hunk 1 above)
Comment (baijiaruo) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:373]: fixed

File: curvefs/src/client/s3/client_s3_cache_manager.cpp (hunk 2 above)
Comment (baijiaruo) [reply curvefs/src/client/s3/client_s3_cache_manager.cpp:386]: fixed

**Comment by baijiaruo** [comment]
recheck

**Comment by baijiaruo** [comment]
recheck

**Review by wu-hanqing — APPROVED** [review]

File: curvefs/conf/client.conf [hunk 9]
```diff
@@ -159,7 +159,11 @@ s3.throttle.bpsTotalMB=0
 s3.throttle.bpsReadMB=0
 s3.throttle.bpsWriteMB=0
 s3.useVirtualAddressing=false
-
+# The interval between read failures and retries will become larger and larger,
+# and when the max is reached, retry will be performed at a fixed time.
+s3.maxReadRetryIntervalNum = 10
```
Comment (ilixiaocui) [inline curvefs/conf/client.conf:164]: s3.maxReadRetryNum = 10

**Review by ilixiaocui — APPROVED** [review]

**Pull request merged**